Dump endpoint can cause Error

Utilizing the `/dump` endpoint without specifying the path or using `?path=` with no defined path, not even `/`, causes the query to show up in the `/info` endpoint indefinitely (until next restart).

Error from log:
```
[qtp248483935-103819] WARN org.eclipse.jetty.server.HttpChannel - SERVER:8080/dump?path=
java.lang.AssertionError: Absolute path required
        at org.apache.hadoop.hdfs.server.namenode.INode.getPathNames(INode.java:761)
        at org.apache.hadoop.hdfs.server.namenode.INode.getPathComponents(INode.java:740)
        at org.apache.hadoop.hdfs.server.namenode.FSDirectory.getINodesInPath(FSDirectory.java:1595)
        at org.apache.hadoop.hdfs.server.namenode.FSDirectory.getINode(FSDirectory.java:1602)
        at org.apache.hadoop.hdfs.server.namenode.FSDirectory.getINode(FSDirectory.java:1609)
        at org.apache.hadoop.hdfs.server.namenode.VersionContext.dumpINodeInDetail(VersionContext.java:61)
        at org.apache.hadoop.hdfs.server.namenode.NameNodeLoader.dumpINodeInDetail(NameNodeLoader.java:212)
        at org.apache.hadoop.hdfs.server.namenode.analytics.WebServerMain.lambda$init$21(WebServerMain.java:696)
...<skipping>...
        at java.lang.Thread.run(Thread.java:745)
```

--- a/src/main/java/org/apache/hadoop/hdfs/server/namenode/NameNodeLoader.java
+++ b/src/main/java/org/apache/hadoop/hdfs/server/namenode/NameNodeLoader.java
@@ -209,6 +209,10 @@ public class NameNodeLoader {
    * @throws IOException error in dumping inode info
    */
   public void dumpINodeInDetail(String path, HttpServletResponse resp) throws IOException {
+    if (path == null || path.isEmpty() || !path.startsWith("/")) {
+      throw new IllegalArgumentException(
+          "Please specify non-empty and absolute `path` parameter for /dump.");
+    }
     versionLoader.dumpINodeInDetail(path, resp);
   }
 

--- a/src/test/java/org/apache/hadoop/hdfs/server/namenode/analytics/TestNNAnalyticsBase.java
+++ b/src/test/java/org/apache/hadoop/hdfs/server/namenode/analytics/TestNNAnalyticsBase.java
@@ -134,6 +134,30 @@ public abstract class TestNNAnalyticsBase {
   }
 
   @Test
+  public void testBadDump1() throws IOException {
+    HttpGet get = new HttpGet("http://localhost:4567/dump?path=");
+    HttpResponse res = client.execute(hostPort, get);
+    assertThat(res.getStatusLine().getStatusCode(), is(500));
+    IOUtils.readLines(res.getEntity().getContent()).clear();
+    HttpGet verify = new HttpGet("http://localhost:4567/info");
+    HttpResponse verifyRes = client.execute(hostPort, verify);
+    List<String> output = IOUtils.readLines(verifyRes.getEntity().getContent());
+    assertThat(output, hasItem("Current active connections: 1"));
+  }
+
+  @Test
+  public void testBadDump2() throws IOException {
+    HttpGet get = new HttpGet("http://localhost:4567/dump?path=bad");
+    HttpResponse res = client.execute(hostPort, get);
+    assertThat(res.getStatusLine().getStatusCode(), is(500));
+    IOUtils.readLines(res.getEntity().getContent()).clear();
+    HttpGet verify = new HttpGet("http://localhost:4567/info");
+    HttpResponse verifyRes = client.execute(hostPort, verify);
+    List<String> output = IOUtils.readLines(verifyRes.getEntity().getContent());
+    assertThat(output, hasItem("Current active connections: 1"));
+  }
+
+  @Test
   public void testTop() throws IOException {
     HttpGet get = new HttpGet("http://localhost:4567/top");
     HttpResponse res = client.execute(hostPort, get);

--- a/src/test/java/org/apache/hadoop/hdfs/server/namenode/analytics/TestNNAnalyticsBase.java
+++ b/src/test/java/org/apache/hadoop/hdfs/server/namenode/analytics/TestNNAnalyticsBase.java
@@ -142,7 +142,7 @@ public abstract class TestNNAnalyticsBase {
     HttpGet verify = new HttpGet("http://localhost:4567/info");
     HttpResponse verifyRes = client.execute(hostPort, verify);
     List<String> output = IOUtils.readLines(verifyRes.getEntity().getContent());
-    assertThat(output, hasItem("Current active connections: 1"));
+    assertThat(output, not(hasItem("/dump?path=:default_unsecured_user")));
   }
 
   @Test
@@ -154,7 +154,7 @@ public abstract class TestNNAnalyticsBase {
     HttpGet verify = new HttpGet("http://localhost:4567/info");
     HttpResponse verifyRes = client.execute(hostPort, verify);
     List<String> output = IOUtils.readLines(verifyRes.getEntity().getContent());
-    assertThat(output, hasItem("Current active connections: 1"));
+    assertThat(output, not(hasItem("/dump?path=bad:default_unsecured_user")));
   }
 
   @Test

